src/natcap/invest/spec.py

```diff
@@ -0,0 +1,2227 @@
+
+    validate_spatial_overlap: bool = True
+    """If True, validation will check that the bounding boxes of all
+    top-level spatial inputs overlap (after reprojecting all to the same
+    coordinate reference system)."""
+
+    different_projections_ok: bool = True
+    """Whether spatial inputs are allowed to have different projections. If
+    False, validation will check that all top-level spatial inputs have the
+    same projection. This is only considered if ``validate_spatial_overlap``
+    is ``True``."""
+
+    aliases: set = dataclasses.field(default_factory=set)
+    """Optional. A set of alternative names by which the model can be called
+    from the invest command line interface, in addition to the ``model_id``."""
+
+    def __post_init__(self):
+        self.inputs_dict = {_input.id: _input for _input in self.inputs}
+        self.outputs_dict = {_output.id: _output for _output in self.outputs}
+
+    def get_input(self, key):
+        """Get an Input of this model by its key."""
+        return self.inputs_dict[key]
+
+    def to_json(self):
+        """Serialize an MODEL_SPEC dict to a JSON string.
+
+        Args:
+            spec (dict): An invest model's MODEL_SPEC.
+
+        Raises:
+            TypeError if any object type within the spec is not handled by
+            json.dumps or by the fallback serializer.
+
+        Returns:
+            JSON String
+        """
+
+        def fallback_serializer(obj):
+            """Serialize objects that are otherwise not JSON serializeable."""
+            if isinstance(obj, pint.Unit):
+                return format_unit(obj)
+            # Sets are present in 'geometry_types' attributes of some args
+            # We don't need to worry about deserializing back to a set/array
+            # so casting to string is okay.
+            elif isinstance(obj, set):
+                return str(obj)
+            elif isinstance(obj, types.FunctionType):
+                return str(obj)
+            elif dataclasses.is_dataclass(obj):
+                as_dict = dataclasses.asdict(obj)
+                if hasattr(obj, 'type'):
+                    as_dict['type'] = obj.type
+                return as_dict
+            elif isinstance(obj, IterableWithDotAccess):
+                return obj.to_json()
+            elif obj is int:
+                return 'integer'
+            elif obj is float:
+                return 'number'
+            raise TypeError(f'fallback serializer is missing for {type(obj)}')
+
+        spec_dict = self.__dict__.copy()
+        # rename 'inputs' to 'args' to stay consistent with the old api
+        spec_dict.pop('inputs')
+        spec_dict.pop('inputs_dict')
+        spec_dict.pop('outputs_dict')
+        spec_dict['args'] = self.inputs_dict
+        spec_dict['outputs'] = self.outputs_dict
+        return json.dumps(spec_dict, default=fallback_serializer, ensure_ascii=False)
+
+
+def build_model_spec(model_spec):
```

Comment:
Docstring?

Reply:
I'll add a short docstring! This function (and `build_input_spec` and `build_output_spec`) are just temporary as we transition over from the model spec dictionaries, so I didn't put as much effort into documenting them.